File: sdk/js/src/Asset.js

// Plantquest Js Asset

class Asset {
  def
  data


  constructor(sdk,data) {
    this.sdk = ()=>sdk
    this.def = {
      name: 'asset'
    }
    // data is optional
    this.data = data
  }


  async handleResult(op, res, spec, handler) {
    const status = res.status

    if(200 === status) {
      const json = await res.json()
      // TODO: error
      return handler(json)
    }
    else {
      throw new Error('HTTP-ERROR: '+op+': asset: '+status)
    }
  }


  async save(data) {
    const op = 'save'
    this.data = data
    // TODO: validate data

    const spec = this.sdk().fetchSpec(op,this)
    const res = await this.sdk().options.fetch(spec.url,spec)

    return this.handleResult(op, res, spec, (json)=>{
      this.data = json
      return this
    })
  }


  async load(data) {
    const op = 'load'
    this.data = data
    // TODO: check data.id defined
    // TODO: separate data and query

    const spec = this.sdk().fetchSpec(op,this)
    const res = await this.sdk().options.fetch(spec.url,spec)

   return this.handleResult(op, res, spec, (json)=>{
      this.data = json
      return this
    })
  }


  async remove(data) {
    const op = 'remove'
    this.data = data
    // TODO: check data.id defined

    const spec = this.sdk().fetchSpec(op,this)
    const res = await this.sdk().options.fetch(spec.url,spec)

    return this.handleResult(op, res, spec, (json)=>{
      this.data = json
      return null
    })
  }


  async list(query) {
    const op = 'list'
    // TODO: use query if defined

    const spec = this.sdk().fetchSpec(op,this)
    const res = await this.sdk().options.fetch(spec.url,spec)

    return this.handleResult(op, res, spec, (json)=>{
      return json.list.map(data=>this.sdk().Asset(data))
    })
  }

}


module.exports = {
  Asset
}

File: sdk/ruby/lib/plantquest_sdk/asset.rb

# Plantquest Ruby Asset

require 'json'
require 'net/http'

class Asset
  attr_accessor :sdk, :def, :data

  def initialize(sdk, def_hash)
    @sdk = sdk
    @def = def_hash
  end

  def save(data)
    self.data = data
    # TODO: validate data

    body = sdk.body('load', self)
    method = sdk.method('load', self)
    url = sdk.endpoint('load', self)
    
    response = sdk.options[:fetch].call(url, method, sdk.options[:apikey], body)
    status = response.code.to_i

    if status == 200
      json = JSON.parse(response.body, symbolize_names: true)
      # TODO: handle errors
      self.data = json[:ent]
      return self
    else
      raise "HTTP-ERROR: save: Asset: #{status}"
    end
  end


  def load(data)
    self.data = data
    # TODO: check if data.id is defined

    body = sdk.body('load', self)
    method = sdk.method('load', self)
    url = sdk.endpoint('load', self)

    response = sdk.options[:fetch].call(url, method, sdk.options[:apikey], body)
    status = response.code.to_i

    if status == 200
      json = JSON.parse(response.body, symbolize_names: true)
      # TODO: handle errors
      self.data = json
    else
      raise "HTTP-ERROR: load: Asset: #{status}"
    end

    self
  end

  def list(data = {})
    self.data = data
    # TODO: check if data.id is defined

    body = sdk.body('load', self)
    method = sdk.method('load', self)
    url = sdk.endpoint('load', self)

    response = sdk.options[:fetch].call(url, method, sdk.options[:apikey], body)
    status = response.code.to_i

    if status == 200
      json = JSON.parse(response.body, symbolize_names: true)
      # TODO: handle errors
      self.data = json
    else
      raise "HTTP-ERROR: load: Asset: #{status}"
    end

    self
  end

end

